src/types.rs

```diff
@@ -24,5 +26,133 @@ pub struct TransactionInput {
 
 #[derive(Serialize, Deserialize)]
 pub struct Response {
-    pub method: Option<Function>,
+    pub method: Option<ResponseMethod>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ResponseMethod {
+    name: String,
+    inputs: Vec<ResponseParam>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct ResponseParam {
+    name: String,
+    internal_type: Option<String>,
+    #[serde(rename(serialize = "type"))]
+    kind: String,
+    value: String,
+}
+
+trait ResponseDisplay {
+    fn display(&self) -> String;
+}
+
+impl ResponseDisplay for ParamType {
+    fn display(&self) -> String {
+        match self {
+            ParamType::String => "string".to_string(),
+            ParamType::Address => "address".to_string(),
+            ParamType::Bytes => "bytes".to_string(),
+            ParamType::Int(n) => format!("int{}", n),
+            ParamType::Uint(n) => format!("uint{}", n),
+            ParamType::Bool => "bool".to_string(),
+            ParamType::Array(t) => format!("{}[]", t.display()),
+            ParamType::FixedBytes(n) => format!("bytes{}", n),
+            ParamType::FixedArray(t, n) => format!("{}[{}]", t.display(), n),
+            ParamType::Tuple(_) => "tuple".to_string(),
+        }
+    }
+}
+
+impl ResponseDisplay for &Vec<u8> {
+    fn display(&self) -> String {
+        "0x".to_string() + &hex::encode(self)
+    }
+}
+
+impl ResponseDisplay for &Vec<Token> {
+    fn display(&self) -> String {
+        self.iter()
+            .map(|x| x.display())
+            .collect::<Vec<String>>()
+            .join(",")
+    }
+}
+
+impl ResponseDisplay for Token {
+    fn display(&self) -> String {
+        match self {
+            Token::String(s) => s.clone(),
+            Token::Address(ad) => format!("{:?}", ad),
+            Token::Bytes(b) | Token::FixedBytes(b) => b.display(),
+            Token::Int(n) | Token::Uint(n) => format!("{:?}", n),
+            Token::Bool(b) => format!("{:?}", b),
+            Token::FixedArray(tokens) | Token::Array(tokens) => {
+                format!("[{}]", tokens.display())
+            }
+            Token::Tuple(tokens) => format!("({})", tokens.display()),
+        }
+    }
+}
+
+impl ResponseMethod {
+    pub fn new(function: Function, data: &[u8]) -> Result<ResponseMethod> {
+        Ok(ResponseMethod {
```

Comment:
you can remove all the clones, but for that you need to "play" with borrow checker a little
first, rust _can_ check partially moved structs, but it isn't the best at it, so if it sees something like
```rust
Foo {
 a: value.a,
 b: value.b(),
}
```
it sees, that you first moving `a` from `value`, and then calling `value.b()`, but you've already moved from `value` and it's in kinda broken state

but it's easy to fix just by calling `b()` first:
```rust
let b = value.b(); // value is correct here
Foo {
 a: value.a, // we can break it down here, nothing wrong
 b,
}
```

So in your code this will look like 
```rust
let inputs = function.inputs.....collect();
Ok(ResponseMethod {
    name: function.name,
    inputs,
})
```
that's minus one clone

then you can call `into_iter` instead of `iter`, but it won't do it because you have first `into_iter`, and then you call the function on the type. It's kinda similar to this code
```rust
let a = value.foo(); // moves value into itself
let b = some_func(value.bar()); // captures value by reference
```
again, it's easy to fix, just switch the order. In your example it will look like

```rust
let decoded = function.decode_input(data)?.into_iter();
let inputs = function
    .inputs
    .into_iter()
    .zip(decoded)
    .map(|(input, token)| ResponseParam::new(input, token))
    .collect();
Ok(ResponseMethod {
    name: function.name,
    inputs,
})
```
and that's how you remove unnecessary clones :)